### PR TITLE
[Core] Add Ring Attention Primitives for Context Parallelism

### PR DESCRIPTION
**Pull request by knitcapcat-amd**
## Purpose

Add Ring Attention primitives as a building block for Prefill Context Parallelism (PCP), delivering the **"Ring-CP style attention backend algorithm"** listed in the PCP roadmap ([RFC #25749](https://github.com/vllm-project/vllm/issues/25749), ref [RFC #26133](https://github.com/vllm-project/vllm/issues/26133)).

Ring Attention partitions the sequence across CP ranks: Q stays local while K/V circulate through a P2P ring. Each step computes a partial attention block and incrementally merges the result via online softmax correction (sigmoid/logsigmoid), enabling communication-computation overlap without the post-hoc all-gather + LSE correction required by the existing DCP approach (`cp_lse_ag_out_rs`).

**This PR provides the communication + computation framework only.** It does not enable PCP end-to-end; PCP integration (KV cache management, ModelRunner token sharding, `supports_pcp=True`) will follow in a subsequent PR. This follows the pattern of [#28718](https://github.com/vllm-project/vllm/pull/28718) (PCP infra merged separately from attention backend) and [#34883](https://github.com/vllm-project/vllm/pull/34883) (A2A comm backend for DCP added as a separate PR).

### New files

| File | Lines | Description |
|------|-------|-------------|
| `vllm/distributed/ring_comm.py` | 174 | Async ring P2P communicator with dedicated CUDA stream, event-based sync, and rank-parity deadlock avoidance) |
| `vllm/v1/attention/ops/ring_attn.py` | 236 | `ring_flash_attn_varlen_func` — Ring Attention for packed variable-length sequences, with online softmax merge, KV packing (halves P2P ops per step), double buffering (zero per-step allocation), and cached CUDA stream. Uses `fa_utils.flash_attn_varlen_func` for platform-agnostic FA dispatch. |
| `tests/distributed/test_ring_attn.py` | 293 | 14 correctness configs × CP={2,4}, covering bidirectional + causal, MHA + GQA, bf16 + fp16, single + multi-request packed batches. Supports both pytest+ray (CI) and standalone torchrun. |

### Key design choices

- **Communication-computation overlap**: `RingComm` runs P2P on a dedicated CUDA stream while FlashAttention runs on the compute stream. Profiled **94% of P2P time hidden behind FA compute** on MI300X at S=8192.
- **KV packing**: K and V are concatenated into a single contiguous buffer before P2P, reducing ops per ring step from 4 to 2.
- **Double buffering**: Two pre-allocated receive buffers alternate across steps, eliminating per-step `torch.empty_like` allocation.
- **Cached CUDA stream**: `hipStreamCreate` costs ~2ms on ROCm; a module-level cache avoids this on every call (TODO: move to backend init in PCP integration PR).
- **Uses `fa_utils` for FA dispatch**: No custom backend selection — platform adaptation (CUDA/ROCm/XPU) is handled by vLLM's existing `fa_utils.flash_attn_varlen_func`.

### Known limitations

- Sequence sharding assumes contiguous partitioning (no stripe/zigzag). Dual-chunk partitioning with per-step position-aware masking is planned for the PCP integration PR.
- Each request's `seq_len` must be divisible by `cp_size`.
- Not compatible with `torch.compile` (Python ring loop + `dist.batch_isend_irecv` + explicit CUDA stream management). This is consistent with existing DCP code which also runs in eager mode.
- Does not manage KV cache (caller's responsibility, same as other ops in `vllm/v1/attention/ops/`).

## Test Plan

```bash
# Correctness — torchrun (standalone, 2 GPU)
torchrun --nproc_per_node=2 tests/distributed/test_ring_attn.py

# Correctness — torchrun (standalone, 4 GPU)
torchrun --nproc_per_node=4 tests/distributed/test_ring_attn.py

# Correctness — pytest (CI mode, requires ray)
pytest tests/distributed/test_ring_attn.py -v
```

## Test Result

Tested on 8× AMD MI300X (ROCm 7.1.1, PyTorch 2.9.1, flash_attn 2.8.3).

**Correctness: CP=2 (14/14 PASS), CP=4 (14/14 PASS)** — all max_diff ≤ 0.003906 at ATOL=5e-3.

```
=== Ring Attention varlen tests (CP=2) ===
  [PASS] single_bidir_MHA       [PASS] multi_eq_bidir        [PASS] multi_diff_bidir
  [PASS] single_causal_MHA      [PASS] multi_eq_causal       [PASS] multi_diff_causal
  [PASS] single_bidir_GQA       [PASS] single_causal_GQA     [PASS] multi_bidir_GQA
  [PASS] multi_causal_GQA       [PASS] multi_bidir_fp16      [PASS] single_GQA_fp16
  [PASS] long_causal_GQA        [PASS] multi_long_causal_GQA
```

**Latency benchmark** (1 request, GQA 32q/8kv heads, D=128, bf16, MI300X):

*Bidirectional (perfectly balanced across ranks):*

| seq_len | FA 1GPU | CP=2 | Speedup | CP=4 | Speedup | CP=8 | Speedup |
|---------|---------|------|---------|------|---------|------|---------|
| 4K | 0.92ms | 0.88ms | 1.04x | 1.20ms | 0.77x | 2.51ms | 0.37x |
| 8K | 3.25ms | 2.25ms | 1.44x | 1.91ms | 1.71x | 2.47ms | 1.32x |
| 16K | 13.81ms | 7.47ms | **1.85x** | 4.90ms | **2.82x** | 3.96ms | **3.51x** |
| 32K | 56.36ms | 29.12ms | **1.94x** | 15.40ms | **3.69x** | 10.87ms | **5.22x** |

*Causal (contiguous partition, load imbalanced):*

| seq_len | FA 1GPU | CP=2 | Speedup | CP=4 | Speedup | CP=8 | Speedup |
|---------|---------|------|---------|------|---------|------|---------|
| 8K | 1.73ms | 1.84ms | 0.94x | 1.75ms | 1.02x | 2.28ms | 0.78x |
| 16K | 6.83ms | 5.73ms | **1.19x** | 4.45ms | **1.54x** | 3.57ms | **1.92x** |
| 32K | 26.76ms | 22.11ms | **1.21x** | 13.48ms | **2.03x** | 9.39ms | **2.91x** |

> Note: Causal results use contiguous partitioning without load balancing. Dual-chunk/zigzag partitioning (per-step position-aware masking, as in TransformerEngine) would improve causal scaling and is planned for the PCP integration PR.

**Overlap profiling** (S=8192, CP=2, `torch.profiler` trace):

```
compute stream:  |███████ step0 FA 4129us ██████|█|███████ step1 FA 3511us ██████|
comm stream:        |████ NCCL P2P 3769us ██████|
                 → 94% of P2P hidden behind FA compute
```

---

<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

**Comment by knitcapcat-amd** [comment]
@LucasWilkinson @MatthewBonanni Hi, could you add the `ready` label to trigger CI? This is my first PR to vllm. Thank you!
